nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp

```diff
@@ -252,41 +252,78 @@ bool BtActionServer<ActionT, NodeT>::loadBehaviorTree(const std::string & bt_xml
-              "ID: %s File: %s",
-              current_bt_id.c_str(), entry.path().string().c_str());
+  std::set<std::string> registered_ids;
+  std::string main_id;
+  auto register_all_bt_files = [&](const std::string & skip_file = "") {
+      try {
+        for (const auto & directory : search_directories_) {
+          for (const auto & entry : fs::directory_iterator(directory)) {
+            if (entry.path().extension() != ".xml") {
+              continue;
+            }
+            if (!skip_file.empty() && entry.path().string() == skip_file) {
+              continue;
+            }
+
+            auto id = bt_->extractBehaviorTreeID(entry.path().string());
+            if (id.empty()) {
+              RCLCPP_ERROR(logger_, "Skipping BT file %s (missing ID)", entry.path().c_str());
+              continue;
+            }
+            if (registered_ids.count(id)) {
+              RCLCPP_WARN(logger_, "Skipping conflicting BT file %s (duplicate ID %s)",
+              entry.path().c_str(), id.c_str());
+              continue;
+            }
+
+            RCLCPP_INFO(logger_, "Registering Tree from File: %s", entry.path().string().c_str());
+            bt_->registerTreeFromFile(entry.path().string());
+            registered_ids.insert(id);
           }
-          bt_->registerTreeFromFile(entry.path().string());
         }
+      } catch (const std::exception & e) {
+        setInternalError(ActionT::Result::FAILED_TO_LOAD_BEHAVIOR_TREE,
+      "Exception reading behavior tree directory: " + std::string(e.what()));
       }
-    } catch (const std::exception & e) {
-      setInternalError(ActionT::Result::FAILED_TO_LOAD_BEHAVIOR_TREE,
-        "Exception reading behavior tree directory: " + std::string(e.what()));
-      return false;
-    }
-  }
-  // Try to load the main BT tree (by ID)
+    };
+
   try {
-    if(!is_bt_id) {
-      tree_ = bt_->createTreeFromFile(file_or_id, blackboard_);
+    if (!is_bt_id) {
+    // file_or_id is a filename: register it first
+      std::string main_file = file_or_id;
+      main_id = bt_->extractBehaviorTreeID(main_file);
+      if (main_id.empty()) {
+        RCLCPP_ERROR(logger_, "Failed to extract ID from %s", main_file.c_str());
+        return false;
+      }
+      RCLCPP_INFO(logger_, "Registering Tree from File: %s", main_file.c_str());
+      bt_->registerTreeFromFile(main_file);
+      registered_ids.insert(main_id);
+
+    // When a filename is specified, it must be register first
+    // and treat it as the "main" tree to execute.
+    // This ensures the requested tree is always available
+    // and prioritized, even if other files in the directory have duplicate IDs.
+    // The lambda then skips this main file to avoid
+    // re-registering it or logging a duplicate warning.
+    // In contrast, when an ID is specified, it's unknown which file is "main"
+    // so all files are registered and conflicts are handled in the lambda.
+      register_all_bt_files(main_file);
```

Comment:
This is now in a try/catch. We don't need a try/catch in `register_all_bt_files` anymore thankfully. That simplifies the lambda